DHT Sensor support

Tannoo opened the pull request:
Adding the option to read the local ambient temp and humidity from a DHT
sensor.

thinkyhead committed:
Marlin: Slight cleanup to Marlin_main.cpp

thinkyhead committed:
Cleanup to planner.cpp

thinkyhead committed:
Marlin: Temperature soft-PWM cleanup

Commit:
Show filament width sensor values in LCD constantly if SD card support is not enabled.

Tannoo committed:
DHT Sensor support

Adding the option to read the local ambient temp and humidity from a DHT
sensor.

Tannoo committed:
Default configuration correction

Travis-Build correction

Blue-Marlin commented on Marlin/Configuration.h:
```diff
@@ -1460,4 +1460,26 @@
+    #define DHT_PIN A3      //Set pin for the DHT sensor. Default = A3 (AUX1 connector pin3 (outside row))
+
+#endif //DHT_ENABLE
+
```
There are much more Configuration.h-files in the `example_configurations` directory. All of them have to be modified.

Maybe `Configuration_adv.h` would be the better place for this feature.

Tannoo replied:
PITA, but done.

Blue-Marlin commented on Marlin/language_en.h:
```diff
@@ -462,6 +462,9 @@
 #endif
+#ifndef MSG_DHT_WARMING
+  #define MSG_DHT_WARMING                     _UxGT("Sensor Warming...")
+#endif
```
Seems to be unused.

Tannoo replied:
Yes...I wanted to display "Sensor Warming..." during the delay. I am still trying to nail down how to create that screen.

Blue-Marlin commented on Marlin/ultralcd.cpp:
```diff
@@ -2389,6 +2396,34 @@ void kill_screen(const char* lcd_msg) {
+          STATIC_ITEM(MSG_INFO_DHT_MENU);                                            //Temp & Humidity
+          STATIC_ITEM("    Temp (C): ", false, true, itostr3left(DHT_temperature));  //Temp (C): 14
+          STATIC_ITEM("    Humidity: ", false, true, itostr3left(DHT_humidity));     //Humidity: 27
+        END_SCREEN();
```
While this strings should be translatable.

Tannoo replied:
I thought about that. But the leading spaces were better alignment than the "center" option did.

Blue-Marlin commented on Marlin/ultralcd.cpp:
```diff
@@ -2389,6 +2396,34 @@ void kill_screen(const char* lcd_msg) {
+          DHT dht;
+          dht.setup(DHT_PIN, DHT_TYPE);
+          safe_delay (1800);
+          DHT_temperature = (dht.getTemperature());
```
'safe_delay' is a good solution here - but the ' ' before the '(' is ugly.
```diff
-          DHT_temperature = (dht.getTemperature());
+          DHT_temperature = dht.getTemperature();
```

Tannoo replied:
Corrected for your viewing pleasure. :)

Tannoo committed:
Moved to configuration_adv

Cleaned up and added to ALL configuration_adv files.

Tannoo committed:
String translations

Tannoo committed:
Visual corrections

Tannoo committed:
Sanity Check addition

Tannoo committed:
Sensor warming screen

Click to start sensor warming.

akaJes commented on Marlin/SanityCheck.h:
```diff
@@ -261,6 +261,15 @@
+  #if DISABLED(ULTIPANEL)
+    #error "DHT Sensor option requires an LCD controller."
+  #endif
+#endif
```
also it need LCD_INFO_MENU enabled

Tannoo replied:
Done.

Tannoo replied:
Should I move this to sanity check instead of automaticly enabling the `lcd_info_menu`?

Roxy-3D replied:
Yes...   Sanity.cpp would be the best place for this check.   But I presume you would be putting a:
```cpp
#if ENABLED(DHT_ENABLE)
```
in front of that check, right?

Tannoo replied:
Of course. :)

Tannoo replied:
There is no `Sanity.cpp`. There is a `SanityCheck.h`.

Tannoo commented on Marlin/SanityCheck.h:
```diff
@@ -261,6 +261,15 @@
+  #if DISABLED(ULTIPANEL)
+    #error "DHT Sensor option requires an LCD controller."
+  #endif
+#endif
```
How about this instead?

`#define DHT_ENABLE`
`#if ENABLED(DHT_ENABLE)`
    `#ifndef LCD_INFO_MENU`
      `#define LCD_INFO_MENU //This must be defined.`
    `#endif`
    `#define DHT_TYPE 22     //DHT type 11, 21, or 22`
    `#define DHT_PIN A3      //Set pin for the DHT sensor.`
`#endif //DHT_ENABLE`

Tannoo replied:
In `configuration_adv.h`.

Tannoo committed:
LCD_INFO_MENU

Enable the LCD_INFO_MENU if DHT_ENABLE is defined.

Bob-the-Kuhn committed:
G38 optional double touch

Made the double touch portion a conditional compile based on the
PROBE_DOUBLE_TOUCH flag.

==============================================

Bugfix

The current G38 only stopped a move if it involved the Z axis.

Moved all the G38 code to it's own section and put it where it would
always be executed no matter what axis was moving or if the endstop was
enabled.

Also added a comment to configuration_adv to alert the user the double
tap had to be turned on.

==============================================

Change G38 back to using Z_MIN_PROBE

There's no Z_MIN endstop if Z_DUAL_ENDSTOPS is enabled and you have them
set to the top of the gantry.

G38 started out as using the Z_MIN_PROBE pin.  I don't remember why we
changed it to the Z_MIN endstop.

Tannoo committed:
Moved LCD_INFO_MENU

Changed LCD Info Menu to be sanity checked instead of automatically
enabled.

Tannoo committed:
Updated to latest RCBugFix

2-15-17 RCBugFix

Tannoo committed:
Merge branch 'RCBugFix' into RCBugFix---DHT

Tannoo committed:
Update Readme

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

StefanBruens committed:
Makefile: remove no longer existing search paths

ArduinoAddons has been removed from Marlin, remove any references to it.
Replace HARDWARE_SRC and HARDWARE_DIR with its expanded equivalents.

Signed-off-by: Stefan Brüns <stefan.bruens@rwth-aachen.de>

StefanBruens committed:
Makefile: add printcounter.cpp too CXXSRC list

The file is required by the optional PRINTCOUNTER feature.

Signed-off-by: Stefan Brüns <stefan.bruens@rwth-aachen.de>

StefanBruens committed:
Makefile: Fix linker error

Instantiation of the static singleton instance thermalManager in general
requires locking, but we know its safe to do without.

Add -fno-threadsafe-statics to avoid using locking primitives, which
result in linking errors:
	undefined reference to `__cxa_guard_acquire'
	undefined reference to `__cxa_guard_release'

Signed-off-by: Stefan Brüns <stefan.bruens@rwth-aachen.de>

MagoKimbra committed:
Fix delta_angles

StefanBruens committed:
Add Support TinyBoy2 OLED/Encoder panel

The OLED is driven by an SSD1306, connected to the board via
I2C, the rotary encoder is connected to 3 GPIO pins.

Signed-off-by: Stefan Brüns <stefan.bruens@rwth-aachen.de>

StefanBruens committed:
Makefile: add nozzle.cpp too CXXSRC list

The file is required by the optional NOZZLE_PARK_FEATURE.

Signed-off-by: Stefan Brüns <stefan.bruens@rwth-aachen.de>

thinkyhead committed:
Add a "manual" option for ABL

thinkyhead committed:
Implement the "manual" option for ABL

Bob-the-Kuhn committed:
Reduce UBL RAM usage by making G26/G29 optional

thinkyhead committed:
Clean up code, remove _now command function

thinkyhead committed:
Have run_probe() return probe Z, not nozzle Z

thinkyhead committed:
MBL fiddle

thinkyhead committed:
Implement LCD_BED_LEVELING for PROBE_MANUALLY

thinkyhead committed:
Board A4JP should be SCOOVO_X9H

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

thinkyhead committed:
Dynamic feedrate on SCARA, converting target mm/s to deg/s

Roxy-3D committed:
Get G29's P1 (Automated Probing) working again.

Incorrect optimizations of data types and ternary operators caused some
issues.

thinkyhead committed:
Various UBL cleanups and bug fixes

thinkyhead committed:
Tweak "can not" => "can't" in RUMBA pins

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

StefanBruens committed:
Add example configuration for TinyBoy L10/L16

The TB2 is a Indiegogo funded 3d printer. It uses a MKS SMELZI V1.0
main board (based on MELZI, i.e. Sanguinololu). I comes as a complete kit,
including display and panel.
Two variants exists, L10 and L16, with a height of 100 and 160 mm.
The heating function of the provided bed can enabled by soldering a
thermistor and some wires to it.

Signed-off-by: Stefan Brüns <stefan.bruens@rwth-aachen.de>

thinkyhead committed:
Further repairs to UBL, comments, spacing

thinkyhead committed:
Disable UBL_G26_MESH_EDITING by default

thinkyhead committed:
More data in UBL class, make it a static class

- Make all `unified_bed_leveling` data/methods static
- Move some UBL-related variables into the class
- Replace `map_[xy]_index_to_bed_location` with `mesh_index_to_[xy]pos`

thinkyhead committed:
Make G26 compatible with G92 and M206

thinkyhead committed:
Make MBL a static class, use lookup for index-to-point conversion

thinkyhead committed:
Additional UBL fixes, optimizations

thinkyhead committed:
Apply FIXFLOAT macro

thinkyhead committed:
Within applied to UBL

thinkyhead committed:
Add and apply WITHIN macro

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

thinkyhead committed:
Syntax cleanup

thinkyhead committed:
Update description of PINS_DEBUGGING

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

thinkyhead committed:
Revert some changes to UBL

thinkyhead committed:
Revert behavior of enqueue_and_echo_commands_P

thinkyhead committed:
Fix logic of UBL::fade_scaling_factor_for_z

thinkyhead committed:
Remove extraneous check of UBL_G26_MESH_EDITING

thinkyhead committed:
Minor spelling patch

thinkyhead committed:
Apply const args, clean up find_closest_circle_to_print

thinkyhead committed:
Use PSTR versions of lcd_status... to save SRAM

thinkyhead committed:
status_printf => lcd_status_printf_P

thinkyhead committed:
Indent MarlinSerial code

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

thinkyhead committed:
Relocate serial macros and functions

Tannoo committed:
Power Supply 0 as "General"

thinkyhead committed:
Patch MarlinSerial to match up with 32-bit version

thinkyhead committed:
can wstring be removed?

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

thinkyhead committed:
Add .gitattributes to normalize line endings

thinkyhead committed:
Fix formatting, order for a UBL change

thinkyhead committed:
Prevent a warning on echo_command

thinkyhead committed:
Fix LCD preheat menus

thinkyhead committed:
Configurable delay for BLTouch servo emulation

thinkyhead committed:
Fix: Thermal runaway if nonexistent bed's temp is set

thinkyhead committed:
Fix line-endings in UBL_G29

thinkyhead committed:
Patch .gitattributes till UBL line endings are fixed

Roxy-3D committed:
Fix problem described in issue # 6175

As it turns out...   an unsigned int can not go to less than 0,  so the
loop never terminates.

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

Roxy-3D committed:
Fix M421 AUTO_BED_LEVELING_BILINEAR and AUTO_BED_LEVELING_UBL

M421 was not connected up for AUTO_BED_LEVELING_BILINEAR.
M421 needed to migrate mesh data to new UBL EEPROM layout.

Roxy-3D committed:
Fix spelling mistake when UBL's mesh is Inactive

Tannoo committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into RCBugFix---DHT

Tannoo committed:
Merge remote-tracking branch 'origin/RCBugFix---DHT' into RCBugFix---DHT

thinkyhead commented:
Looks like this PR got lost in the shuffle while I was being homeless. Are you planning to re-submit?
